File: src/main/java/me/felix/tired/api/Tired.java
package me.felix.tired.api;

import lombok.Getter;
import me.felix.tired.bridge.Managers;
import me.felix.tired.bridge.event.Listener;
import me.felix.tired.bridge.managers.ShaderManager;
import me.felix.tired.modification.Module;

import java.util.ArrayList;

@Getter
public class Tired {

    @Getter
    private static Tired tired;

    private final Managers managers = new Managers();
    private final ArrayList<Plugin> plugins = new ArrayList<>();
    @Getter
    private final ShaderManager shaderManager = new ShaderManager();
    public Tired() {
        tired = this;
    }

    public ArrayList<Module> getModules() {
        return tired.managers.getModuleManager().getModules();
    }

    public static void addListener(Class<? extends Listener> listener) {
        if(!tired.managers.getEventManager().eventClasses.contains(listener))
            tired.managers.getEventManager().eventClasses.add(listener);
    }

    public static void removeListener(Class<? extends Listener> listener) {
        tired.managers.getEventManager().eventClasses.remove(listener);
    }
}

File: src/main/kotlin/tired/jdk/api/Tired.kt
package tired.jdk.api

import tired.jdk.api.event.Listener
import tired.jdk.api.abstracts.Module
import tired.jdk.intern.Plugin

object Tired {

    private var listeners = ArrayList<Class<out Listener>>()
    private val modules = ArrayList<Module>()
    private val plugins = ArrayList<Plugin>()

    fun addModule(module: Module) {
        modules.add(module)
    }

    fun getModules() : ArrayList<Module> {
        return modules
    }

    fun <T : Module?> getModule(clazz: Class<T>?): T {
        return modules.stream().filter { module: Module ->
            module.javaClass == clazz
        }.findAny().orElse(null) as T
    }

    fun <T : Module?> getModule(name: String?): T {
        return modules.stream().filter { module: Module ->
            module.name.equals(name, true)
        }.findAny().orElse(null) as T
    }

    fun getListeners() : ArrayList<Class<out Listener>> {
        return listeners
    }

    fun addListener(listener: Class<out Listener>) {
        if(!listeners.contains(listener))
            listeners.add(listener)
    }

    fun removeListener(listener: Class<out Listener>) {
        listeners.remove(listener)
    }
}